x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java

```diff
@@ -552,16 +600,18 @@ void getOrCreateProfileWithBackoff(
                 );
                 updateProfileForActivate(subject, versionedDocument, listener);
             } else {
-                // The profile document is NOT a match, this means either genuine hash collision or profile document
-                // was manually updated, e.g. realm or user rename etc. So we attempt to differentiate from the
-                // existing profile document by increase the differentiator number by 1.
-                incrementDifferentiatorAndCreateNewProfile(subject, profileDocument, listener);
+                // The profile document is NOT a match, this means either:
+                // 1. Genuine hash collision
+                // 2. A different user has the same username
+                // 3. Profile document was manually updated
+                // So we attempt to differentiate from the existing profile document by increase the differentiator number by 1.
+                maybeIncrementDifferentiatorAndCreateNewProfile(subject, profileDocument, listener);
             }
         }, listener::onFailure));
     }
 
     // Package private for tests
-    void incrementDifferentiatorAndCreateNewProfile(Subject subject, ProfileDocument profileDocument, ActionListener<Profile> listener)
+    void maybeIncrementDifferentiatorAndCreateNewProfile(Subject subject, ProfileDocument profileDocument, ActionListener<Profile> listener)
         throws IOException {
         final String uid = profileDocument.uid();
         final int index = uid.lastIndexOf("_");
```

Comment:
I think this is a problem because the uid suffix can contain "_".

Reply:
Good point! I think the simplest thing to do is to not allow it. I will tighten it to be just alphanumeric.